### DISCUSSION TITLE
FFlate `zip` callback not functioning!

### DISCUSSION THREAD

**Discussion by Umisyus** [101arrowz/fflate#155] [Q&A] [answered]
Hi there, I'm trying to use this `zip` function to return my zipped files out of my function through variable assignment. The problem is, when I do that the callback does not assign my result and I receive undefined.
Here is an example in Typescript:

```ts
function processFile(left: number, file: { key: string, value: Buffer }) {
    let result: Uint8Array | undefined;

    bufToU8(file.value, (buf: Uint8Array) => {
        // With fflate, we can choose which files we want to compress
        zipObj[file.key] = [buf, {
            level: 0
        }];

        // If we didn't want to specify options:
        // zipObj[file.name] = buf;
    });
    if (!--left) {
       zip(zipObj, function (err, out) {
            if (err) { console.error(err); throw err; }

/* PROBLEM HAPPENS HERE */
            result = out
            console.log("Done zipping");

/* THIS WORKS, BUT I NEED A WAY TO RETURN FROM THE `processFile` FUNCTION */
// fs.writeFile('test.zip', out)

        });
       
    }

    return result
}
```

Thanks for your time.

**Comment by 101arrowz** [comment, marked as answer]
The problem is that `zip` is an asynchronous function, but you want to use it in a synchronous function. Asynchronous functions can only be used with callbacks or (preferably) `Promise`. In other words, what you're asking is not possible; you should probably create a callback parameter for `processFile` to "return" your value.

Also, I'm not sure the `--left` bit of this code will work if you call `processFile` multiple times, as it doesn't actually decrement the value in the caller. `left` is passed by value to `processFile`, not as a reference.

**Reply by Umisyus** [reply]
Thanks for your quick reply. So `zip` is an async function, so I could do this:
```
const files = [
{key:'test1',value:"ABCDEFGH123"},
{key:'test2',value:"ABCDEFGH12345"},
]

const myZipOptions = {level: 0}
let myObjectWithAFileTree = {}

for (let file of files) { myObjectWithAFileTree[file.key] = [ file.value ] }

export const zipFiles = (
  data: AsyncZippable,
  options: AsyncZipOptions = { level: 0 }
): Promise<Uint8Array> => {
  return new Promise((resolve, reject) => {
    zip(data, options, (err, data) => {
      console.warn("err = ", err);
      console.log("data = ", data);
      if (err) return reject(err);
      return resolve(data);
    });
  });
};
```


Then I would call it like this: await zipFiles(myObjectWithAFileTree, myZipOptions)
Regarding the left parameter, I have it decremented outside of the function.
```
let files = [
{key:'test1',value:"ABCDEFGH123"},
{key:'test2',value:"ABCDEFGH12345"},
]

let myObjectWithAFileTree = {}
for (let file of files) { myObjectWithAFileTree[file.key] = [ file.value ] }

let zipResult = await zipFiles(myObjectWithAFileTree,myZipOptions)

fs.promises.writeFile("results.zip", zipResult)
```


Regarding the left parameter, I have it decremented outside of the function.

**Reply by 101arrowz** [reply]
Yes, your first example is nearly correct. However, if you have strings as your values, you need to make them Uint8Array to use. Like this:

```js
import { strToU8 } from 'fflate';
// Instead of: for (let file of files) { myObjectWithAFileTree[file.key] = [ file.value ] }
for (let file of files) { myObjectWithAFileTree[file.key] = [ strToU8(file.value) ];
```

**Reply by Umisyus** [reply]
So I should be good then? Thanks a lot man. My values are Buffers, which use `UInt8Array`s.

Thanks a lot man.
Peace

**Comment by Umisyus** [comment]
Issue may be closed now.
